frontend/src/hooks/useValidation.js

```diff
@@ -0,0 +1,341 @@
+    const sorted = orderBy(parsedRanges, ["startDate"]);
+
+    // Combine overlapping ranges
+    const consolidated = sorted.reduce((merged, current) => {
+      const lastRange = merged.at(-1);
+
+      // If the start date of the current range is before the end date of the last range,
+      // combine the ranges
+      if (lastRange && isBefore(current.startDate, lastRange.endDate)) {
+        lastRange.endDate = max([lastRange.endDate, current.endDate]);
+      } else {
+        merged.push(current);
+      }
+
+      return merged;
+    }, []);
+
+    return consolidated;
+  }
+
+  // Returns true if all reservation date ranges are within the operating date ranges
+  function validateReservationDates(dateableFeature) {
+    // Consolidate any overlapping date ranges for comparison
+    const operatingRanges = consolidateRanges(dateableFeature.Operation);
+    const reservationRanges = consolidateRanges(dateableFeature.Reservation);
+
+    // Check if each reservation ranges is within an operating range
+    const withinRange = reservationRanges.every((reservationRange) =>
+      operatingRanges.some((operatingRange) =>
+        checkWithinRange(
+          operatingRange.startDate,
+          operatingRange.endDate,
+          reservationRange.startDate,
+          reservationRange.endDate,
+        ),
+      ),
+    );
+
+    return withinRange;
+  }
+
+  // Validates the start and end datepicker fields in a date range
+  function validateDateRange({
+    dateRange,
+    start,
+    startDateId,
+    end,
+    endDateId,
+    datesObj = dates,
+  }) {
+    clearError(startDateId);
+    clearError(endDateId);
+
+    // If both dates are blank, ignore the range
+    if (!start && !end) {
+      return true;
+    }
+
+    // Both dates are required if one is set
+    if (!start) {
+      return addError(startDateId, "Enter a start date");
+    }
+
+    if (!end) {
+      return addError(endDateId, "Enter an end date");
+    }
+
+    // Parse date strings
+    const startDate = new Date(start);
+    const endDate = new Date(end);
+
+    // Check if the start date is before the end date
+    if (startDate > endDate) {
+      return addError(
+        endDateId,
+        "Enter an end date that comes after the start date",
+      );
+    }
+
+    const operatingYear = season.operatingYear;
+
+    // Date must be within the year for that form
+    if (
+      startDate.getFullYear() !== operatingYear ||
+      endDate.getFullYear() !== operatingYear
+    ) {
+      // startDate =< endDate check happens first, so the end date will never fail this check
+      return addError(startDateId, `Enter dates for ${operatingYear} only`);
+    }
+
+    const dateType = dateRange.dateType.name;
+    const { dateableId } = dateRange;
+
+    // Check if the reservation dates are all within the operating dates
+    const dateableFeature = datesObj[dateableId];
+    const reservationDatesValid = validateReservationDates(dateableFeature);
```

Comment:
Would it make more sense to call this function inside of validateFeatureDates only once?
Right now, validateFeatureDates calls validateDateRange for every dateRange in the feature and `validateReservationDates` would be called for every dateRange even though is basically validating the same thing for the whole feature.

Reply:
There's a fair bit of unnecessary repeat validating happening right now... I'd like to maybe make it just validate the entire form (or the entire campsite grouping section, at least) once on every change, since so many rules depend on each other. It would un-tangle the code a fair bit, and then we could easily debounce it if we want. I'm gonna wait to hear from UX about the updates to the error displays. If we only need to display an error at the feature level, then it makes sense to just check errors at that level too. (We wouldn't need to track the ID of the field that changed)

I agree with you, though! I'll check with Amanda and see if that's something we'll be doing in this ticket, or a bit later.